File: language/java_study/javaLearn/src/main/java/com/iamhotcenplotz/chapter13/Exercise.java
package com.iamhotcenplotz.chapter13;

/**
 * @author Der Hotcenplotz
 * @version 1.0
 */
public class Exercise {
    public static void main(String[] args) {
        System.out.println(Exercise01.reverse("000 123 000", 3, 7));
    }
}

class Exercise01 {
    public static String reverse(String str, int start, int end) {
        if(start >= end){
            throw new RuntimeException("...");
        }

        if (end > str.length()){
            throw new RuntimeException("+++");
        }


        char[] charArray = str.toCharArray();

        char temp;

        for (int i = start, j = end; i < j; i++, j--) {
            temp = charArray[i];
            charArray[i] = charArray[j];
            charArray[j] = temp;
        }

        return String.valueOf(charArray);
    }
}

File: language/scala_study/ScalaLearn/src/main/scala/com/hotcenplotz/chapter11/Exercise_.scala
package com.hotcenplotz.chapter11

import scala.collection.mutable.ArrayBuffer

/**
 * @author: Der Hotcenplotz
 * @version： 1.0 
 * @date: 2025-05-19 20:32 
 * @description:
 * */
object Exercise_ {
    
    def main(args: Array[String]): Unit = {
        exercise02()
    }
    
    def exercise01():Unit = {
        val str = "AAAAAAAAAAAAAAABBBBBBBBCCCCCCCCCCCCCDDDDDDDDD"
        val arrbuf: ArrayBuffer[Char] = new ArrayBuffer[Char]()
        
        def temp(arr:ArrayBuffer[Char],c:Char):ArrayBuffer[Char] = {
            arr.append(c)
            arr
        }
        
        str.foldLeft(arrbuf)(temp)
        
        println(arrbuf)
    }
    
    def exercise02():Unit = {
        val str = "AAAAAAAAAAAAAAABBBBBBBBCCCCCCCCCCCCCDDDDDDDDD"

        def temp(m:Map[Char,Int],c:Char):Map[Char,Int] = {
            m + (c -> (m.getOrElse(c,0) + 1))
        }
        
        println(str.foldLeft(Map[Char, Int]())(temp))
    }
    
}
